_source/_posts/2023-01-23-keycloak-kubernetes-prod.md

```diff
@@ -0,0 +1,644 @@
+  config {
+    baseName store
+    reactive true
+    packageName com.okta.developer.store
+    applicationType microservice
+    authenticationType oauth2
+    buildTool gradle
+    databaseType mongodb
+    enableHibernateCache false
+    serverPort 8082
+    serviceDiscoveryType consul
+  }
+  entities Product
+}
+
+entity Blog {
+  name String required minlength(3)
+  handle String required minlength(2)
+}
+
+entity Post {
+  title String required
+  content TextBlob required
+  date Instant required
+}
+
+entity Tag {
+  name String required minlength(2)
+}
+
+entity Product {
+  title String required
+  price BigDecimal required min(0)
+  image ImageBlob
+}
+
+relationship ManyToOne {
+  Blog{user(login)} to User with builtInEntity
+  Post{blog(name)} to Blog
+}
+
+relationship ManyToMany {
+  Post{tag(name)} to Tag{post}
+}
+
+paginate Post, Tag with infinite-scroll
+paginate Product with pagination
+
+microservice Product with store
+microservice Blog, Post, Tag with blog
+
+deployment {
+  deploymentType docker-compose
+  appsFolders [gateway, blog, store]
+  dockerRepositoryName "indiepopart"
+}
+```
+
+Repalce the `dockerRepositoryName` with your account name. Then run the following command:
+
+```shell
+jhipster jdl reactive-ms.jdl
+```
+
+After the generation, you will find sub-folders were created for the `gateway`, `store`, and `blog` services. The `gateway` will act as the front-end application and as a secure router to the `store` and `blog` microservices.
+
+Create a [Docker Hub](https://hub.docker.com/) personal account, if you don't have one. Then build and publish each application image. For example, in the `gateway` folder run:
+
+```shell
+./gradlew -Pprod bootJar jib \
+-Djib.to.image=<your-dockerhub-username>/gateway \
+-Djib.to.auth.username=<your-dockerhub-username> \
+-Djib.to.auth.password=<your-dockerhub-secret>
+```
+
+**Note**: _your-dockerhub-secret_ can be your Docker Hub password, or a token if you have two-factor authentication enabled.
+
+### Run the Kubernetes sub-generator
+
+Now let's generate the Kubernetes descriptors using JHipster. During the process, the generator will prompt for an FQDN (Fully qualified domain name) for the Ingress services, so let's first create a public IP on Google Cloud. With the help of [nip.io](nip.io), if you set <public-ip>.nip.io as your FQDN, you can test the deployment without having to purchase a real domain.
+
+Google Cloud provides a [free tier](https://cloud.google.com/free) of their services that grants you $300 in free credits if you are a new user.
+
+After you sign up, install [`gcloud` CLI](https://cloud.google.com/sdk/docs/install). When you reach the end of the process, the last step is to run `gcloud init` and set up authorization for the tool. Also, install the [Kubectl authentication plugin](https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke).
+
+```shell
+gcloud compute addresses create gateway-ip --global
+```
```

Comment:
When I first ran this in a new project, it prompted me to install compute.

```
API [compute.googleapis.com] not enabled on project [379941542179]. Would you like to enable and retry (this will take a
 few minutes)? (y/N)?  y
```

Not sure if we should mention this or not.

Reply:
I will QA with a clean project and add notes.

Reply:
Added instructions to enabled the required APIs in advance.